### ISSUE TITLE
Custom PIPL Module

### ISSUE THREAD

**Issue by cyberlive-debug** [smicallef/spiderfoot#1077]
I was wondering about getting the pipl functionality into Spiderfoot.  I am using Spiderfoot 3.2 and python 3.8.  

I have created some custom modules using similar tools such as PeopleData and SocialSearcher but the difference is they will run fine with one type element being sent in.  PIPL works the best when it has many items sent in, the more you put in the more you get back.  So, for a robust return of data if i was to run it manually, I would add email, name, address, username, etc.  My understanding with Spiderfoot is it looks for certain events in the watched events, and when they happen, it grabs it.  Because of this, my assumption is that if I trapped for email, name, address, username, I would not be able to create a query string containing all the 4 elements, but it would call PIPL 4 times with 1 element each.

Is that correct?

If so, is there a way for me to build a string to create a more robust request for pipl to send four elements once rather then one element four times?

If more clarification is needed, please let me know.

Thanks for the support.

**Comment by bcoles** [comment]
> My understanding with Spiderfoot is it looks for certain events in the watched events, and when they happen, it grabs it. Because of this, my assumption is that if I trapped for email, name, address, username, I would not be able to create a query string containing all the 4 elements, but it would call PIPL 4 times with 1 element each.
> 
> Is that correct?

That is correct. The current model uses event notifications, using one event of one type per notification.

Similarly, modules do not have access to events previously identified from other modules. You could hack together some module code which queries the database from within the Plugin context, but I wouldn't recommend it. This breaks the model and will likely have unforeseen consequences.

> 
> If so, is there a way for me to build a string to create a more robust request for pipl to send four elements once rather then one element four times?

There is no way to do this with the current model.

I don't see this changing any time soon either. (Perhaps Steve will weigh in to the contrary.)

Unfortunately, the event pub/sub model is both SpiderFoot's greatest strength and biggest weakness.

**Comment by bcoles** [comment]
Here's a slightly longer answer, with a couple of alternatives, that will probably only confuse you.

---

You could create an event type in the database schema and a module which consumes this event type. (or use an existing event type such as `RAW_RIR_DATA`, and use custom parsing within the module), and use this as a scan seed target. (this is easier than it sounds).

This would require formatting your data as a single event data string (ie, CSV or JSON formatting). This would also require modifying SpiderFoot to accept your new input type as a scan target. This would allow you to run a SpiderFoot scan with your specified inputs as a single event, which would be fed into your custom Pipl module, which would then pass the data on to Pipl.

This approach is hackish and pretty far from ideal. Taking this approach removes many advantages of running SpiderFoot. That is, no identified data (apart from your seed target) will ever be fed to the Pipl module, but data identified from the Pipl module can be fed into other modules.

---

Another option would be to write your own wrapper for Pipl, external to SpiderFoot, and have your custom module consume your wrapper as a data source. Unfortunately, this probably wouldn't work and would require a fair bit of development effort to setup. SpideFoot does not properly support module priority, thus your custom module may be executed before all four of your desired data types are identified. Also, I'm not familiar enough with Pipl to know whether writing a wrapper would be feasible. This depends on your use case, but probably wouldn't be feasible, as you would need to identify which email address belongs to which human name belongs to which username, etc.

**Comment by smicallef** [comment]
This is indeed a limitation and something I have an idea around solving. The general concept would be around having "signal" events that inform modules when a scan is in a certain state (e.g. starting, no events left, paused, completely finished). On such events, modules can respond in whatever way they please. In your example, the module would listen for all the events of interest, keep them in memory, and when signaled that there are no more events to process, it can generate the query to PIPL.

There are some limitations in this approach and open questions, but it's not a top priority to solve so I wouldn't expect it any time soon. We can keep this bug open for reference and as a reminder.
